src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Performance/CSharpDoNotGuardSetAddOrRemoveByContains.Fixer.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System.Composition;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.NetCore.Analyzers.Performance;
+
+namespace Microsoft.NetCore.CSharp.Analyzers.Performance
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp), Shared]
+    public sealed class CSharpDoNotGuardSetAddOrRemoveByContainsFixer : DoNotGuardSetAddOrRemoveByContainsFixer
+    {
+        protected override bool SyntaxSupportedByFixer(SyntaxNode conditionalSyntax, SyntaxNode childStatementSyntax)
+        {
+            if (childStatementSyntax is not ExpressionStatementSyntax)
+            {
+                return false;
+            }
+
+            if (conditionalSyntax is IfStatementSyntax ifStatementSyntax)
+            {
+                return ifStatementSyntax.Statement.ChildNodes().Count() == 1;
+            }
+
+            return false;
+        }
+
+        protected override Document ReplaceConditionWithChild(Document document, SyntaxNode root, SyntaxNode conditionalOperationNode, SyntaxNode childOperationNode)
+        {
+            SyntaxNode newRoot;
+
+            if (conditionalOperationNode is ConditionalExpressionSyntax conditionalExpressionSyntax &&
+                conditionalExpressionSyntax.WhenFalse.ChildNodes().Any())
+            {
+                var expression = GetNegatedExpression(document, childOperationNode);
+
+                SyntaxNode newConditionalOperationNode = conditionalExpressionSyntax
+                    .WithCondition((ExpressionSyntax)expression)
+                    .WithWhenTrue(conditionalExpressionSyntax.WhenFalse)
+                    .WithWhenFalse(null!)
```

Comment:
❗ This is going to throw ArgumentNullException

Reply:
Good catch, I thought this is for removing the false part, did not notice the '!'. What can we use instead?

Reply:
I originally copied this section from the similar fixer for dictionaries (`CSharpDoNotGuardDictionaryRemoveByContainsKeyFixer`).
As part of a bugfix for the dictionary analyzer, it was discovered that the case of the ternary operator was never flagged by the analyzer (see https://github.com/dotnet/roslyn-analyzers/pull/6770#discussion_r1264964323).
So the code in the fixer that would throw an ArgumentNullException is never executed.

That part should probably be removed from the dictionary fixer as well (I missed that in the PR above):
https://github.com/dotnet/roslyn-analyzers/blob/47edf21cd38fb583da54085bdfd724c2cdf1f6f0/src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Performance/CSharpDoNotGuardDictionaryRemoveByContainsKeyFixer.cs#L44-L48

Reply:
Thanks for explanation @mpidash, removing this part sounds good to me, what you think @sharwell?

Reply:
I believe if this is reachable, it would would be hit by:

```
bool added = set.Contains(value) ? false : set.Add(value);
```

Reply:
I've added two tests for ternary operators:

1. One with a ternary operator as in your example. In this case, no diagnostic is reported.
2. Another test with a nested ternary operator. No diagnostic is reported for the `Add` case, but one is reported for the `Remove` case.

This is because we only check the true part of a conditional for an applicable `Add` or `Remove`. So the `Add` case gets filtered out immediately because it contains no child operations:

```C#
var firstOperation = operations.FirstOrDefault();

if (firstOperation is null)
{
    return false;
}
```

For `Remove` it depends on the content of the true part. In the simple case, it contains a `FieldReferenceOperation`, which is not covered by the switch in `HasApplicableAddOrRemoveMethod` so it produces no diagnostic.
The nested case for `Remove` produces a diagnostic because the first child operation is a `IInvocationOperation`.

A ternary operator gets filtered out by the fixer in `SyntaxSupportedByFixer`, because the `childStatementSyntax` is not a `ExpressionStatementSyntax` (it is `ConditionalExpressionSyntax`). So the part where the `ArgumentNullException` would be thrown is not reachable.

I think we should filter out ternary operators for now. I am not sure how likely it is that someone would write code like in the test cases, as they already use the return value of `Add` and `Remove`.
`DoNotGuardDictionaryRemoveByContainsKey` should be adapted to treat ternary operators in the same way as this analyzer.

---

The above has shown a case that is not covered at the moment:

```C#
if (MySet.Contains(""Item""))
{
    throw new Exception(""Item already exists"");
}
else
{
    MySet.Add(""Item"");
}
```

which could be refactored to:

```C#
if (!MySet.Add(""Item""))
{
    throw new Exception(""Item already exists"");
}
```

Reply:
> I think we should filter out ternary operators for now. I am not sure how likely it is that someone would write code like in the test cases, as they already use the return value of Add and Remove.
> DoNotGuardDictionaryRemoveByContainsKey should be adapted to treat ternary operators in the same way as this analyzer.

Agree that it would be rare, though it could happen and it fixable. I am OK with filtering this out for now and create an issue for future fix.

> The above has shown a case that is not covered at the moment:

Did you find why it was not covered and could you fix this with the PR?

Reply:
> Did you find why it was not covered and could you fix this with the PR?

Its because atm only the true branch of a condition is checked for `Add` and `Remove`.
I will add a fix which also checks the false branch.

Reply:
- [ad82982](https://github.com/dotnet/roslyn-analyzers/pull/6767/commits/ad8298285efd463cd5ac557387ff2ddc35267225) adds support for `Add` and `Remove` in else blocks
- [f723ac5](https://github.com/dotnet/roslyn-analyzers/pull/6767/commits/f723ac580879390b3ef4762bee690c4594555caf) adds support for ternary diagnostics (no fixer)

Reply:
Thank you for the fix @mpidash!

@sharwell requested change is addressed PTAL